### ISSUE TITLE
T-online.de webmail renders conditional comments

### ISSUE THREAD

**Issue by hteumeuleu** [hteumeuleu/email-bugs#43]
The desktop webmail of [T-online.de](https://www.t-online.de/) content within HTML conditional comments. So the following example:

```html
<!-- <p>This is an HTML comment.</p> -->
<!--[if mso]>
<p>This is a mso conditional comment.</p>
<![endif]-->
<!--[if foobar]>
<p>This is a foobar conditional comment.</p>
<![endif]-->
```
…is transformed by T-online.de into the following:

```html
<p>This is a mso conditional comment.</p>
<p>This is a foobar conditional comment.</p>
```
![A screenshot of T-online.de rendering conditional comments](https://user-images.githubusercontent.com/3451753/40705417-18fb00bc-63eb-11e8-8911-4bd7a836d197.png)

This only affects conditional comments (not regular HTML comments). And this works with any conditional keyword (like `foobar` in the previous example).

Thanks to @M-J-Robbins for spotting this one yesterday on the [#emailgeeks Slack](https://email.geeks.chat/).

**Comment by brunorilla** [comment]
Is there any solution for this?

**Comment by DarkDust** [comment]
There's a workaround we're using in [Mail Designer 365](https://maildesigner365.com): Outlook doesn't understand `display:none;`¹⁾ but T-Online does. So you can do:

```
<p>Renders everywhere.</p>
<!--[if mso]>
<p style="display:none;">Renders in Outlook only.</p>
<![endif]-->
```

1) AFAIK, Outlook only understands `display:none;` on TABLE elements.

**Comment by M-J-Robbins** [comment]
@DarkDust ah that's clever worth adding to the list here if you can https://howtotarget.email/

Also to target just T-Online you can use `<!--[if false]> T-Online <![endif]-->`

**Comment by longtowner** [comment]
new Outlook Clients support `display: none`, that's why we need an additional `<!--[if mso]>p.target-class {display: block !important}<![endif]-->` to show the content, if it's hidden. t-online keeps the content hidden, because it deletes all the `<style>` Blocks

**Comment by adrianthomas** [comment]
> new Outlook Clients support `display: none`, that's why we need an additional `<!--[if mso]>p.target-class {display: block !important}<![endif]-->` to show the content, if it's hidden. t-online keeps the content hidden, because it deletes all the `<style>` Blocks

You mean they support `display: none` outside of tables? Which clients, the new iOS Outlook?

**Comment by longtowner** [comment]
@adrianthomas yes Outlook 365 supports it at least for <center> elements. i tried it today. I had to hide the Text inside a VML Button (https://buttons.cm/) because T-Online shows the Button Text twice.

**Comment by calebrauxa** [comment]
Thanks so much for documenting this and making my life easier today.

**Comment by SpecOps12** [comment]
> 

@longtowner You made my day! Works like a charm 🪄

**Comment by NivenRanchhod** [comment]
It looks like this issue has been rectified. No mso content is being rendered, as far as I can see 🎉

However, a separate issue I'm now seeing is that style blocks are being stripped from the head and body and T-Online is expecting inlined CSS 🤦🏾

**Comment by M-J-Robbins** [comment]
Nice, thanks for sharing @NivenRanchhod this is great news

**Comment by eabarrio1** [comment]
Hi all, it appears this issue is happening again, as of November 7th. 

In our case, it is happening with this instance, which is causing our email header to disappear. 


Should appear on any client that's not Outlook on Windows


Anyone else seeing this, or know when it may be resolved?

**Comment by eabarrio1** [comment]
You can report the issue to the provider here: https://www.telekom.de/kontakt/e-mail-kontakt/unternehmen/ihre-telekom-idee?fs_id=31beb80a-f487-45bd-8aff-b38380762ba0. I just put 00000000000 as the callback number, as it requires that field.

**Comment by mariusglauer** [comment]
@eabarrio1 we have the same issue since a few days on our side, can confirm.

**Comment by husseinalhammad** [comment]
It seems to remove all HTML comments, which is reasonable. What it doesn't do correctly is handling the `if !mso` (not Microsoft Office conditional comments).

## `if mso` comments (if MSO)

It's ok if T-Online removes this. We only want it for Outlook on Windows (classic) anyway.

```html
<!--[if mso]>
  This is only displayed on Outlook Windows (classic) desktop apps
<![endif]-->
```

## `if !mso` comments (if not MSO)

Expectation:

```html
<!--[if !mso]><!-->
  I am only hidden on Outlook Windows (classic) desktop apps
<!--<![endif]-->
```

Actual results:

```html
<!--[if !mso]><!-->
  I am hidden on Outlook Windows (classic) desktop apps & T-Online
<!--<![endif]-->
```

So T-Online removes the whole thing including the content sandwiched in between the opening and closing comments, which is not what we expect.

### A workaround

```html
<span style="display:none;"><!--[if !mso]><! fix -> --></span>
  I am only hidden on Outlook Windows (classic) desktop apps
<!--<![endif]-->
```

1. `<!--[if !mso]><! fix -> -->` takes advantage of how T-Online incorrectly parses HTML comments so the opening comment gets removed, but what comes after it (the content) is kept
2. The opening comment is wrapped with `<span style="display:none;"></span>` to workaround a Yahoo bug https://github.com/hteumeuleu/email-bugs/issues/95. Without it, `-->` is rendered.
3. The closing comment can be kept as is. It gets removed regardless.

**Comment by eabarrio1** [comment]
My coworker found a solution for the if !mso comments (if not MSO). 

`<!--[if !mso]><!-->`
  I am only hidden on Outlook Windows (classic) desktop apps
`<!--<![endif]-->`

Add spaces to the opening and closing comments:

`<!--[if !mso]><! -->`
  I am only hidden on Outlook Windows (classic) desktop apps
`<!-- <![endif]-->`

Make sure that all comments have spaces as well:  should be 

This will keep the hidden content where it's needed!

**Comment by NivenRanchhod** [comment]
Looks to be working for me, with no adverse effects in Outlook clients 🎉🎉🎉
